template/template.go

```diff
@@ -0,0 +1,203 @@
+package template
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	ast "go.expect.digital/mf2/parse"
+)
+
+type execFn func(operand any, opts map[string]any) (string, error)
+
+type Template struct {
+	ast       *ast.AST
+	execFuncs map[string]execFn
+	executer  *executer
+}
+
+type executer struct {
+	wr    io.Writer
+	input map[string]any
+}
+
+func (e *executer) writeString(s string) error {
+	if _, err := e.wr.Write([]byte(s)); err != nil {
+		return fmt.Errorf("write: %w", err)
+	}
+
+	return nil
+}
+
+func newExecuter(wr io.Writer, input map[string]any) *executer {
+	return &executer{wr: wr, input: input}
+}
+
+func New() *Template {
+	return &Template{execFuncs: make(map[string]execFn)}
+}
+
+func Must(t *Template, err error) *Template {
+	if err != nil {
+		panic(err)
+	}
+
+	return t
+}
+
+// AddFunc adds a function to the template's function map.
+func (t *Template) AddFunc(name string, f execFn) {
```

Comment:
Please use exposed types in the exposed function. Always.

?

```suggestion
func (t *Template) AddFunc(name string, f Func) {
```

Reply:
yeah, makes sense